---
layout: post
title: CGIからのパーミッション
date: 2018-08-10 03:55:07
categories: apache gpio cgi
---
<p>Webから操作するリモコンカメラを作ろうとしています。</p>

<p>左右回転するモータドライバをGPIOでON、OFFさせる<code>motor_rotate.sh</code>をお手軽にbashで作り<br>
所有者：グループをroot:rootとし、SUIDを付与しました。</p>

```
-rwsr-xr-x 1 root  root   898 Aug  9 10:25 motor_rotate.sh
```

<p>このスクリプトをuser権限で実行できることを確認しました。</p>

<p>次に、<code>motor_rotate.sh</code>を呼び出すCGI、<code>cam_ctl.py</code>をPythonで作りました。</p>

```
-rwxr-xr-x 1 root root 190 Aug 10 07:08 cam_ctl.py
```

<p>しかし、ブラウザからこのCGIを実行すると以下のエラーになり、</p>

```
The server encountered an internal error or misconfiguration and was unable to complete your request.
```

<p>apache2/error.logは次のようになります。</p>

```
/home/user1/bin/motor_rotate.sh: 20: /home/user1/bin/motor_rotate.sh: cannot create /sys/class/gpio/export: Permission denied
/home/user1/bin/motor_rotate.sh: 22: /home/user1/bin/motor_rotate.sh: cannot create /sys/class/gpio/gpio23/direction: Permission denied
/home/user1/bin/motor_rotate.sh: 23: /home/user1/bin/motor_rotate.sh: cannot create /sys/class/gpio/gpio23/value: Permission denied
/home/user1/bin/motor_rotate.sh: 34: /home/user1/bin/motor_rotate.sh: cannot create /sys/class/gpio/gpio23/value: Permission denied
/home/user1/bin/motor_rotate.sh: 38: /home/user1/bin/motor_rotate.sh: cannot create /sys/class/gpio/gpio23/value: Permission denied
[Fri Aug 10 07:41:09.833310 2018] [cgid:error] [pid 18429:tid 2985292848] [client 192.168.12.150:50335]
 End of script output before headers: cam_ctl.py, referer: http://192.168.12.103:80/
```

<p>どのように対処したらよいでしょうか？<br>
そもそも、CGIの実行者は誰になっているのでしょうか？<br>
よろしくお願いいたします。</p>

<p>ちなみに、<code>motor_rotate.sh</code>はこんな感じです（略してますので、行番号は違います）</p>

```
#!/bin/sh
SWITCH1="/sys/class/gpio/gpio23"

if [ ! -e $SWITCH1 ]; then
    echo 23 &gt; /sys/class/gpio/export
    sleepenh 0.1 &gt; /dev/null
    echo out &gt; $SWITCH1/direction
    echo 0 &gt; $SWITCH1/value
fi

echo 1 &gt; $SWITCH1/value
sleepenh 0.5 &gt; /dev/null
echo 0 &gt; $SWITCH1/value
```

<p>追伸、AMD64版Debinaでテストしてみたんですが<br>
-rwxr-xr-x  1 user1  user1     73  8月 10 17:01 test.py<br>
-rwsr-xr-x  1 root   root      30  8月 10 17:00 test.sh</p>

<p>test.sh</p>

```
#! /bin/sh
echo "Hello World"
```

<p>test.sh</p>

```
#! /usr/bin/python
import subprocess
res = subprocess.call("./test.sh")
```

<p>user1@debian:~$ ./test.py<br>
Hello World</p>

<p>このテストでは、問題いなく動くんです、何か間違ってるのかな</p>
